File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        self.summ = float("-inf")

        def dfs(root):
            if not root:
                return 0

            left = dfs(root.left)
            right = dfs(root.right)
            leftMax = max(left, 0)
            rightMax = max(right, 0)

            current_sum = root.val + leftMax + rightMax 
            self.summ = max(self.summ, current_sum)

            return root.val + max(leftMax, rightMax)

        dfs(root)
        return self.summ

File: Data Structures & Algorithms/house-robber-iii/submission-1.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */
 
func rob(root *TreeNode) int {
	var check func(root *TreeNode) (int, int)

	check = func(node *TreeNode) (int, int) {
		if node == nil {
			return 0, 0
		}

		leftRob, leftSkip := check(node.Left)
		rightRob, rightSkip := check(node.Right)
		
		robThis := node.Val + leftSkip + rightSkip
		skipThis := max(leftRob, leftSkip) + max(rightRob, rightSkip)
		
		return robThis, skipThis
	}

	return max(check(root))
}
